external-providers/java-external-provider/pkg/java_external_provider/provider.go

```diff
@@ -966,3 +997,68 @@ func (p *javaProvider) BuildSettingsFile(m2CacheDir string) (settingsFile string
 
 	return settingsFilePath, nil
 }
+
+func getJavaExecutable(validateJavaVersion bool) (string, error) {
+	javaExecutable := "java"
+	if javaHome, exists := os.LookupEnv("JAVA_HOME"); exists {
+		javaExecToTest := filepath.Join(javaHome, "bin", "java")
+		if runtime.GOOS == "windows" {
+			javaExecToTest += ".exe"
+		}
+		if _, err := os.Stat(javaExecToTest); err == nil {
+			javaExecutable = javaExecToTest
+		}
+	}
+
+	if !validateJavaVersion {
+		return javaExecutable, nil
+	}
+
+	out, err := exec.Command(javaExecutable, "-version").CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("failed to run %s -version: %w", javaExecutable, err)
+	}
+
+	re := regexp.MustCompile(`version\s"(\d+)[.\d]*"`)
+	matches := re.FindStringSubmatch(string(out))
+	if len(matches) > 1 {
+		javaVersion := matches[1]
+		if majorVersion := javaVersion; majorVersion < "17" {
+			return "", errors.New("jdtls requires at least Java 17")
+		}
+		return javaExecutable, nil
+	}
+
+	return "", errors.New("could not determine Java version")
+}
+
+func findEquinoxLauncher(jdtlsBaseDir string) (string, error) {
+	pluginsDir := filepath.Join(jdtlsBaseDir, "plugins")
+	files, err := os.ReadDir(pluginsDir)
+	if err != nil {
+		return "", fmt.Errorf("failed to read plugins directory: %w", err)
+	}
+
+	for _, file := range files {
+		if strings.HasPrefix(file.Name(), "org.eclipse.equinox.launcher_") && strings.HasSuffix(file.Name(), ".jar") {
+			return filepath.Join(pluginsDir, file.Name()), nil
+		}
+	}
+
+	return "", errors.New("cannot find equinox launcher")
+}
+
+func getSharedConfigPath(jdtlsBaseDir string) (string, error) {
+	var configDir string
+	switch runtime.GOOS {
+	case "linux", "freebsd":
+		configDir = "config_linux"
+	case "darwin":
+		configDir = "config_mac"
+	case "windows":
+		configDir = "config_win"
+	default:
+		return "", fmt.Errorf("unknown platform %s detected", runtime.GOOS)
+	}
+	return filepath.Join(jdtlsBaseDir, configDir), nil
+}
```

Comment:
The file is getting quite big, we'll probably have to think about refactoring at some point :sweat_smile: